### ISSUE TITLE
XML schema errors on payment submission

### ISSUE THREAD

**Issue by jeffwhitfield** [thephpleague/omnipay-authorizenet#38]
I'm currently working on a SilverStripe Omnipay implementation (https://github.com/silverstripe/silverstripe-omnipay) and I'm running into an error with the Authorize.net Omnipay module. 

To start, upon testing submissions, I got the following error: 

> Warning] simplexml_load_string(): namespace warning : xmlns: URI AnetApi/xml/v1/schema/AnetApiSchema.xsd is not absolute

After digging a little, I updated `AIMResponse.php` on line 27 with the following: 

` $xml = simplexml_load_string($xml, 'SimpleXMLElement', LIBXML_NOWARNING);`

This got rid of the error but revealed another problem. Here's the value of the $xml variable, which is a SimpleXMLElement Object response from the call to Authorize.net:

```xml
SimpleXMLElement Object (
  [messages] => SimpleXMLElement Object (
    [resultCode] => Error
    [message] => SimpleXMLElement Object (
      [code] => E00003
      [text] => The element 'transactionRequest' in namespace 'AnetApi/xml/v1/schema/AnetApiSchema.xsd' has invalid child element 'billTo' in namespace 'AnetApi/xml/v1/schema/AnetApiSchema.xsd'. List of possible elements expected: 'cardholderAuthentication, retail, employeeId, transactionSettings, userFields' in namespace 'AnetApi/xml/v1/schema/AnetApiSchema.xsd'.
    )
  )
)
```

And here is the SimpleXMLElement Object value from the getData() method in `AIMAuthorizeRequest.php`:

```xml
SimpleXMLElement Object (
  [@attributes] => Array (
    [xmlns] => AnetApi/xml/v1/schema/AnetApiSchema.xsd
  )
  [merchantAuthentication] => SimpleXMLElement Object (
    [name] => SimpleXMLElement Object ( )
    [transactionKey] => SimpleXMLElement Object ( )
  )
  [refId] => 14-14
  [transactionRequest] => SimpleXMLElement Object (
    [transactionType] => authCaptureTransaction
    [amount] => 125.00
    [payment] => SimpleXMLElement Object (
      [creditCard] => SimpleXMLElement Object (
        [cardNumber] => 4111111111111111
        [expirationDate] => 0120
        [cardCode] => 123
      )
    )
    [customerIP] => 192.168.0.1
    [billTo] => SimpleXMLElement Object (
      [firstName] => John
      [lastName] => Doe
      [company] => SimpleXMLElement Object ( )
      [address] => 5555 Main Street
      [city] => New York
      [state] => New York
      [zip] => 10001
      [country] => US
    )
    [shipTo] => SimpleXMLElement Object (
      [firstName] => John
      [lastName] => Doe
      [company] => SimpleXMLElement Object ( )
      [address] => 5555 Main Street
      [city] => New York
      [state] => New York
      [zip] => 10001
      [country] => US
    )
    [transactionSettings] => SimpleXMLElement Object (
      [setting] => SimpleXMLElement Object (
        [settingName] => testRequest
        [settingValue] => false
      )
    )
  )
)
```

Don't see anything out of the ordinary so...I'm confused. Any idea on what could be causing these errors?  Thx in advance!  :)

**Comment by judgej** [comment]
Well, it certainly isn't absoulte:

~~~json
  [@attributes] => Array (
    [xmlns] => AnetApi/xml/v1/schema/AnetApiSchema.xsd <<<<<<
  )
~~~

The question is, where is it coming from? Can you give us any of the code that is generating this request?

**Comment by judgej** [comment]
Some info:

http://www.scriptarticle.com/get-rid-of-warning-in-authorized-net-transaction-details-api/ (fixes the Authorize.Net SDK, but does not explain how it works or why)

https://community.developer.authorize.net/t5/Integration-and-Testing/namespace-warning-xmlns-URI-AnetApi-xml-v1-schema-AnetApiSchema/td-p/51649 (suppressing the error)

https://github.com/thephpleague/omnipay-authorizenet/blob/bd9523cbc2bbe74e1caf297ffb2a5effbb583c75/src/Message/AIMAbstractRequest.php#L165

I suspect you have already seen these, but I'm just catching up :-)

The suppression of the warning is *probably* the correct approach, and needs to go into this package. The next error is more about the structure of the data sent:

> The element 'transactionRequest' in namespace 'AnetApi/xml/v1/schema/AnetApiSchema.xsd' has invalid child element 'billTo' in namespace 'AnetApi/xml/v1/schema/AnetApiSchema.xsd'. List of possible elements expected: 'cardholderAuthentication, retail, employeeId, transactionSettings, userFields' in namespace 'AnetApi/xml/v1/schema/AnetApiSchema.xsd'.

**Comment by judgej** [comment]
`billTo` and `shipTo` look fine to me too - [for AIM](www.authorize.net/content/dam/authorize/documents/AIM_guide_XML.pdf) - so I don't understand that error. Are you using AIM or CIM here?

**Comment by judgej** [comment]
Someone here with a similar issue:

https://community.developer.authorize.net/t5/Integration-and-Testing/Bug-in-AIM-over-XML-is-lt-order-gt-part-of-the-schema-or-not/td-p/16860

It seems the *order* of the nodes in the XML is important. Out-of-order elements will result in this error. `customerIP` must come *AFTER* the `billTo` and `shipTo` elements. Could you try it out at your end without setting the customer IP at all, and see if that makes suppresses this error?

**Comment by judgej** [comment]
If that works, this line moved down one (move `addCustomerIp()` below `addBillingData()`) is likely to be the permanent fix:

https://github.com/thephpleague/omnipay-authorizenet/blob/bd9523cbc2bbe74e1caf297ffb2a5effbb583c75/src/Message/AIMAuthorizeRequest.php#L20

There may even be a `addShippingData()` call missing from there too - not set up here to try that out right now.

**Comment by judgej** [comment]
I've created a new branch we can use to try out these changes. I've made two two we have so far:

https://github.com/thephpleague/omnipay-authorizenet/tree/issue38

**Comment by judgej** [comment]
@jeffwhitfield Have you managed to give this a try yet?

**Comment by judgej** [comment]
I've managed to demonstrate the error on the omnipay-example application against the AIM sandbox using the latest release of this gateway, getting this response:

~~~php
object(SimpleXMLElement)#172 (1) {
  ["messages"]=>
  object(SimpleXMLElement)#395 (2) {
    ["resultCode"]=>
    string(5) "Error"
    ["message"]=>
    object(SimpleXMLElement)#394 (2) {
      ["code"]=>
      string(6) "E00003"
      ["text"]=>
      string(350) "The element 'transactionRequest' in namespace 'AnetApi/xml/v1/schema/AnetApiSchema.xsd' has invalid child element 'billTo' in namespace 'AnetApi/xml/v1/schema/AnetApiSchema.xsd'. List of possible elements expected: 'cardholderAuthentication, retail, employeeId, transactionSettings, userFields' in namespace 'AnetApi/xml/v1/schema/AnetApiSchema.xsd'."
    }
  }
}
~~~

Switching to the `issue38` branch gives me no errors and a successful authorisation:

~~~php
object(SimpleXMLElement)#172 (3) {
  ["refId"]=>
  string(8) "09876786"
  ["messages"]=>
  object(SimpleXMLElement)#396 (2) {
    ["resultCode"]=>
    string(2) "Ok"
    ["message"]=>
    object(SimpleXMLElement)#395 (2) {
      ["code"]=>
      string(6) "I00001"
      ["text"]=>
      string(11) "Successful."
    }
  }
  ["transactionResponse"]=>
  object(SimpleXMLElement)#394 (12) {
    ["responseCode"]=>
    string(1) "1"
    ["authCode"]=>
    string(6) "000000"
    ["avsResultCode"]=>
    string(1) "P"
    ["cvvResultCode"]=>
    object(SimpleXMLElement)#395 (0) {
    }
    ["cavvResultCode"]=>
    object(SimpleXMLElement)#397 (0) {
    }
    ["transId"]=>
    string(1) "0"
    ["refTransID"]=>
    object(SimpleXMLElement)#398 (0) {
    }
    ["transHash"]=>
    string(32) "7C44D083D32575C0A3ADA409ED2E3AF8"
    ["testRequest"]=>
    string(1) "1"
    ["accountNumber"]=>
    string(8) "XXXX0027"
    ["accountType"]=>
    string(4) "Visa"
    ["messages"]=>
    object(SimpleXMLElement)#399 (1) {
      ["message"]=>
      object(SimpleXMLElement)#400 (2) {
        ["code"]=>
        string(1) "1"
        ["description"]=>
        string(35) "This transaction has been approved."
      }
    }
  }
}
~~~

Being actually a simple change (and fixing a potentially show-stopper issue), I'm going merge in issue38 and make a patch release, just to get it out there and hopefully save some pain.

**Comment by judgej** [comment]
Released as 2.4.1
